universal-hash/src/lib.rs

```diff
@@ -0,0 +1,167 @@
+//! Traits for [Universal Hash Functions].
+//!
+//! Universal hash functions select from a "universal family" of possible
+//! hash functions selected by a key. They are well suited to the purpose
+//! of "one time authenticators" for a sequence of bytestring inputs,
+//! as their construction has a number of desirable properties such as
+//! pairwise independence as well as amenability to efficient implementations,
+//! particularly when implemented using SIMD instructions.
+//!
+//! When combined with a cipher, such as in Galois/Counter Mode or the
+//! Salsa20 family AEAD constructions, they can provide the core functionality
+//! for a Message Authentication Code (MAC).
+
+#![no_std]
+#![doc(html_logo_url = "https://raw.githubusercontent.com/RustCrypto/meta/master/logo_small.png")]
+#![warn(missing_docs)]
+
+#[cfg(feature = "std")]
+extern crate std;
+
+pub extern crate generic_array;
+extern crate subtle;
+
+use generic_array::{ArrayLength, GenericArray};
+use subtle::{Choice, ConstantTimeEq};
+
+/// The `UniversalHash` trait defines a generic interface for universal hash
+/// functions.
+pub trait UniversalHash<BlockSize: ArrayLength<u8>>: Clone
```

Comment:
Shouldn't be `BlockSize` an associated type instead?

Reply:
If you can figure out how to make the bound on the next line work with associated types...

```rust
where
    BlockSize::ArrayType: Copy,
```

I don't think Rust's type system supports associated type bounds yet:

https://github.com/rust-lang/rust/issues/52662

Reply:
IIUC you need the `Copy` bound only for de-referncing in `*GenericArray::from_slice(chunk)`. In that case I think we can just replace it with clone (see the other thread).